Source/WebGPU/WebGPU/Sampler.mm

```diff
@@ -272,41 +289,62 @@ static uint64_t floatToUint64(float f)
     return !!m_samplerIdentifier;
 }
 
+NSUInteger Sampler::aliveCount()
+{
+    Locker locker { samplerStateLock };
+    [weakSamplerStates _markNeedsCompaction];
+    [weakSamplerStates compact];
+    return weakSamplerStates.count;
+}
+
 id<MTLSamplerState> Sampler::samplerState() const
 {
     if (!m_samplerIdentifier)
         return nil;
 
-    Locker locker { samplerStateLock };
-    if (!cachedSamplerStates) {
-        cachedSamplerStates = [NSMutableDictionary dictionary];
-        lastAccessedKeys = [NSMutableOrderedSet orderedSet];
+    {
+        Locker locker { samplerStateLock };
+        if (!cachedSamplerStates) {
+            cachedSamplerStates = [NSMutableDictionary dictionary];
+            lastAccessedKeys = [NSMutableOrderedSet orderedSet];
+            weakSamplerStates = [NSPointerArray weakObjectsPointerArray];
+        }
+
+        id<MTLSamplerState> samplerState = [cachedSamplerStates objectForKey:m_samplerIdentifier];
+        if (samplerState) {
+            m_cachedSamplerState = samplerState;
+            return samplerState;
+        }
     }
 
-    id<MTLSamplerState> samplerState = [cachedSamplerStates objectForKey:m_samplerIdentifier];
-    if (samplerState)
-        return samplerState;
-
     id<MTLDevice> device = m_device->device();
     if (!device)
         return nil;
-    if (cachedSamplerStates.count >= device.maxArgumentBufferSamplerCount) {
+
+    if (aliveCount() >= device.maxArgumentBufferSamplerCount) {
```

Comment:
The device is used by CA and WebGL too. I don't think you can use the max amount of samplers, other code breaks.

Reply:
CA and WebGL don't set `supportArgumentBuffers=YES` https://developer.apple.com/documentation/metal/mtlsamplerdescriptor/2915782-supportargumentbuffers?language=objc

Indeed if they started doing so, we would need to adopt a new strategy, like swizzling `+[_MTLSampler load]` or something